### DISCUSSION TITLE
Make websocket handlers async

### DISCUSSION THREAD

**Discussion by protortyp** [actix/actix-web#3063] [Ideas] [answered]
Hey everyone,
while implementing much of our logic with actix-web and its websocket handlers, I've noticed an issue. Much of what we do with databases (sqlx), redis, kafka and so on, needs some sort of async operations. And right now, this is quite a pain when we're dealing with websocket handlers. We have to spawn new threads or wrap our futures in `block_on` handlers that complicate the logical flow.

Seeing that Rust now has stable async traits (🎉🎉) got me wondering if there are any plans to get async support into websocket handlers? It would certainly make our lives much easier. 

I really like having a combined REST-API with websockets endpoints in one go. And all the [extras](https://github.com/actix/actix-extras) like protobufs support, would make a switch to an architecture based on actix-web for REST + tokio-tungstenite for websockets, a bit of a pain.

Keen to hear your thoughts!

**Comment by asonix** [comment, marked as answer]
If you're using actix's actor-based websocket handler, you can use async methods, and you shouldn't use block_on in an actor context.

https://docs.rs/actix/latest/actix/type.ResponseFuture.html
https://docs.rs/actix/latest/actix/type.ResponseActFuture.html

**Comment by robjtede** [comment]
(copying from Discord for future wanderers)

There's the `actix-ws` crate i'd recommend having a go with: https://crates.io/crates/actix-ws
It's much easier to do async/await with this crate and we're going to introduce a similar API in Actix Web itself at some point in v4.x.

the `*-actorless` examples are equivalent demonstrations using actix-ws: https://github.com/actix/examples/tree/master/websockets

That combined with https://ryhl.io/blog/actors-with-tokio/ will probably be a better system than using actix actors.
